src/v/archival/tests/archiver_operations_impl_gtest.cc

```diff
@@ -0,0 +1,1275 @@
+      .delta_offset = model::offset_delta(0),
+      .archiver_term = expected_archiver_term,
+      .segment_term = expected_segment_term,
+      .delta_offset_end = model::offset_delta(1),
+    });
+    m.advance_applied_offset(expected_applied_offset);
+    m.advance_highest_producer_id(expected_producer_id);
+    m.advance_insync_offset(expected_insync_offset);
+    return m;
+}();
+
+// Condensed version of the record batch header
+struct record_batch_desc_t {
+    model::offset base;
+    model::offset last;
+    model::timestamp ts_base;
+    model::timestamp ts_last;
+    size_t size_bytes;
+    model::record_batch_type type;
+    size_t physical_offset;
+};
+
+struct payload_t {
+    ss::input_stream<char> stream;
+    size_t size;
+    bytes content;
+    std::vector<record_batch_desc_t> batches;
+};
+
+// Generate payload that contains only data batches
+payload_t expected_data_payload(model::offset base, model::offset last) {
+    iobuf payload;
+    std::vector<record_batch_desc_t> batches;
+    for (size_t i = base(); i <= last(); i++) {
+        auto batch = model::test::make_random_batch(model::offset(i), 1, false);
+        auto header_iobuf = storage::batch_header_to_disk_iobuf(batch.header());
+        batches.push_back(record_batch_desc_t{
+          .base = batch.base_offset(),
+          .last = batch.last_offset(),
+          .ts_base = batch.header().first_timestamp,
+          .ts_last = batch.header().max_timestamp,
+          .size_bytes = header_iobuf.size_bytes() + batch.data().size_bytes(),
+          .type = batch.header().type,
+          .physical_offset = payload.size_bytes(),
+        });
+        payload.append(std::move(header_iobuf));
+        payload.append(batch.data().copy());
+    }
+    auto size = payload.size_bytes();
+    auto content = iobuf_to_bytes(payload);
+    auto stream = make_iobuf_input_stream(std::move(payload));
+    //
+    return payload_t{
+      .stream = std::move(stream),
+      .size = size,
+      .content = content,
+      .batches = std::move(batches)};
+}
+
+struct index_payload_t {
+    cloud_storage::segment_record_stats stats;
+    bytes content;
+};
+
+// Generate index payload based on segment content
+index_payload_t expected_index_payload(
+  model::offset_delta initial_delta, std::vector<record_batch_desc_t> bts) {
+    auto filter = raft::offset_translator_batch_types(expected_ntp.to_ntp());
+    index_payload_t result;
+    const auto sampling_step = 64_KiB;
+    auto first = bts.front();
+    offset_index ix(
+      first.base,
+      first.base - initial_delta,
+      initial_delta(),
+      sampling_step,
+      first.ts_base);
+    size_t window = 0;
+    model::offset_delta running_delta = initial_delta;
+    auto it = std::next(bts.begin());
+    for (; it < bts.end(); it++) {
+        auto is_config = [&]() {
+            auto i = std::find(filter.begin(), filter.end(), it->type);
+            return i != filter.end();
+        }();
```

Comment:
nit: do we need an applied lambda here? maybe something like

```
const auto is_config = ranges::find(filter, it->type) != filter.end();
```

would work?

Reply:
done